core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java

```diff
@@ -2508,6 +2512,22 @@ public void testShowCreateDuplicateNames()
                 .hasMessageContaining("Relation 'tpch.s1.duplicate_view' is a view, not a table");
     }
 
+    @Test
+    public void testAnalysisDuplicateNames()
+    {
+        // There is materialized view, view and table with "duplicate_materialized_view" name.
```

Comment:
duplicate_materialized_view is not a mateiralized view (actually: it _is_, but it is _also_ a table and a view)

duplicate_relation?
ambiguous_relation?
overload_relation?

table_view_and_materialized_view_relation?

Reply:
From the engine's point of view, having duplicate view/table/materialized view is a bug, and any behavior around which one wins is undefined and an implementation detail. I don't think we should be codifying this behavior.

Reply:
> From the engine's point of view, having duplicate view/table/materialized view is a bug

I don't think we validate it anywhere. We tried to be consistent in `ShowQueriesRewrite` with how analysis behaves. I improve that in this PR. I think it's good thing to do, but requires testing for previously untested code. What would the alternative be?

Reply:
I agree with @sopel39. We cannot reasonably enforce connectors to implement SPI correctly, so we should make sure Engine doesn't start behaving weirdly when they do.

Having "view takes precedence over table" was already too complex to get right (for example, `HiveMetadata.getTableHandle` returns a TH for a view, even though it shouldn't and it was causing some problems in the past, or so i remember).  Addition pf meterialized views makes things more complex and so we should be exercising those fragile code paths.

Reply:
But then the test is just testing that the code is written the way it's written. The fact that it resolves things in that order is an artifact of how the code is laid out right now and it could change in the future through refactoring. 

I'm fine having a test to exercise behaviors, but what the behavior to become a de-facto standard just because we have a test for it. If we're going to keep the test, let's at least add a comment to the test explaining that it's verifying unspecified behavior that can change at any time, so that if we ever change that behavior, it's clear that that is ok and expected.

Reply:
> But then the test is just testing that the code is written the way it's written.

correct. and it can exercise different code paths from one place too -- ensuring the code for `SELECT t` resolution is consistent with `SHOW CREATE ... t` resolution  -- and these are two different pieces of code

> The fact that it resolves things in that order is an artifact of how the code is laid out right now and it could change in the future through refactoring.

Sounds like yet another reason to have a test.

i am fine with refactor changing the resolution order, as long as it remains consistent across code base.... unless we have a compelling reason to trade off this consistency for something else.

> but what the behavior to become a de-facto standard just because we have a test for it. If we're going to keep the test, let's at least add a comment to the test explaining that it's verifying unspecified behavior that can change at any time

good point.

This should be mentioned in the test (eg as a test method javadoc) and also in the ConnectorMetadata javadoc.

Reply:
Added comment in tests. Changed object names